### DISCUSSION TITLE
Using dag trigger command with parameter

### DISCUSSION THREAD

**Discussion by silverbullet1** [apache/airflow#18601] [Q&A]
Hi everyone,

I am trying to pass some parameters(like schema, table, etc) at runtime using dag trigger --conf command, but I am not sure how I can access them inside my [RedshiftToS3Operator](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_modules/airflow/providers/amazon/aws/transfers/redshift_to_s3.html#RedshiftToS3Operator). From what I read there are 2 ways to access these parameters:

1) Using kwargs (but for that I need to have a PythonOperator will a callable function and then find a way to somehow pass it downstream to Redshift operator, which is unnecessary overhead)

2) Using jijna templating i.e {{ dag_run.conf["param"] }} but its not working with RedshiftToS3Operator

Is there anything else that can be done ?

Thanks!

**Comment by josh-fell** [comment]
How are you using the Jinja templating in the `RedshiftToS3Operator`?  Could you share a code snippet?

**Comment by silverbullet1** [comment]
Hi @josh-fell 
I am using it like this :

unload_task = RedshiftToS3Operator(task_id = "unload_task",
    schema = **"{{ dag_run.conf["schema"] }}"**,
    table = **"{{ dag_run.conf["srcTable"] }}",**
    s3_bucket = Variable.get("s3_bucket"),
    s3_key = Variable.get("s3_key"),
    aws_conn_id = "aws_default",
    redshift_conn_id = **"{{ dag_run.conf["conn_id"] }}"**,
    unload_options = ["FORMAT PARQUET"],
    dag = dag)
    
I have also followed another approach using xcom wherein I get the values using a python operator and use them in my redshift operator, but it doesn't work either. Airflow doesn't render the jinja template and raises **AirflowNotFoundException(f"The conn_id `{conn_id}` isn't defined")** and other variables as well.

Snippet for above attempt :


def init_vars(**kwargs):
    logging.info(kwargs['dag_run'].conf)
    dag_run_conf = kwargs["dag_run"].conf
    kwargs["ti"].xcom_push(key="conn_id", value=dag_run_conf["conn_id"])
    kwargs["ti"].xcom_push(key="schema", value=dag_run_conf["schema"])
    kwargs["ti"].xcom_push(key="srcTable", value=dag_run_conf["srcTable"])

init = PythonOperator(task_id = 'Init_vars', python_callable = init_vars, dag = dag, provide_context = True)

unload_to_s3 = RedshiftToS3Operator(task_id = "unload_to_s3",
    schema = **"{{ task_instance.xcom_pull(key='schema') }}",**
    table =  **"{{ task_instance.xcom_pull(key='srcTable') }}",**
    s3_bucket = Variable.get("s3_bucket"),
    s3_key = Variable.get("s3_key"),
    aws_conn_id = "aws_default",
    redshift_conn_id = **"{{ task_instance.xcom_pull(key='conn_id') }}",**
    unload_options = ["FORMAT PARQUET"],
    dag = dag)

**Reply by josh-fell** [reply]
Unfortunately `redshift_conn_id` isn't a template field for the operator so the `redshift_conn_id` value being passed to the operator would be the literal string "{{ dag_run.conf["conn_id"] }}".  But you could subclass the existing operator to make `redshift_conn_id` a template field for your use case though. 

Is this what you been by "its not working with RedshiftToS3Operator"?  Or is there something else giving you some trouble?

Also, a side note, `s3_bucket` and `s3_key` are currently template fields and you should update the Variable access to Jinja expressions as well.  Otherwise each `Variable.get()` call queries the metadata DB in Airflow for every parsing interval which could lead to some performance degradation in your environment. Something like this would be better:

```python
...
s3_bucket = "{{ var.value.s3_bucket }}",
s3_key = "{{ var.value.s3_key }}",
...
```

**Reply by josh-fell** [reply]
You will also need to change the double-quotes around the `dag_run.conf` key names to single quotes too.

**Reply by silverbullet1** [reply]
Got it, thanks. Thanks a lot for your help. I will try doing that!

**Reply by josh-fell** [reply]
No problem!  Let me know how it goes for you.

**Reply by silverbullet1** [reply]
I tried to run it by hardcoding the connection_id and templating the other fields. It worked if I try to pass **schema = "{{ task_instance.xcom_pull(key='schema') }}"** but did not work with **schema = '{{ dag_run.conf['schema'] }}'**

**Reply by msumit** [reply]
I've tried using `dag_run.conf` only and it worked for me. What Airflow version you are using? Also, have you tried using DAG `params`?

**Reply by silverbullet1** [reply]
I am using v2.1.3. I did not try DAG params. I will read about it.
Anyways, Is there anything wrong if I just push these CLI params to xcom? I think it comes under small data and should not cause any performance issues, even when my jobs are running concurrently?

**Reply by msumit** [reply]
Yeah, there is no issue, it's just that you are forced to take a detour unnecessarily. Try `params`, that should work mostly.

**Reply by silverbullet1** [reply]
Hi @msumit , Is it possible to use the jinja template to access variable value inside a python dictionary like this ?

>      data = {
>              "aws_access_key_id": " {{ var.value.aws_access_key_id }} ",
>              "aws_secret_access_key": "{{ var.value.aws_secret_access_key }}"
>        }

Or do I have to use Variable.get only ?
